File: meudin_app/lib/db/wallet_dao.dart
import 'package:cloud_firestore/cloud_firestore.dart';
import 'package:meudin_app/models/smart_error.dart';
import 'package:meudin_app/models/wallet.dart';
import 'package:meudin_app/services/smart_error_service.dart';

class WalletDao {
  final dbReference = FirebaseFirestore.instance;
  final String _walletsCollectionName = 'wallets';

  late SmartErrorService _smartErrorService;

  WalletDao() {
    _smartErrorService = SmartErrorService();
  }

  // Future<String> updateMemberList(List<String> members, String walletId) async {
  //   var batch = dbReference.batch();

  //   var docRef = dbReference.collection(_walletsCollectionName).doc(walletId);

  //   DateTime lastUpdate = DateTime.now();

  //   batch.set(
  //     docRef,
  //     {
  //       'membersId': members,
  //       'lastUpdate': Timestamp.fromDate(lastUpdate),
  //     },
  //     SetOptions(merge: true),
  //   );

  //   await batch.commit();

  //   return '';
  // }

  // Future<List<Wallet>> getWalletsByUserId(String userId) async {
  //   List<Wallet> wallets = [];

  //   var snapShot = await dbReference
  //       .collection(_walletsCollectionName)
  //       .where('membersId', arrayContains: userId)
  //       .get();

  //   for (var item in snapShot.docs) {
  //     Wallet wallet = Wallet(item);

  //     wallets.add(wallet);
  //   }

  //   wallets.sort((a, b) => a.name.compareTo(b.name));
  //   // }).catchError((onError) {
  //   //   SmartError errorDto = SmartError();
  //   //   errorDto.errorData = onError;
  //   //   errorDto.errorLog = onError.toString();
  //   //   errorDto.feature = 'Get wallets by user id';
  //   //   errorDto.userId = userId;
  //   //   errorDto.userMail = '';

  //   //   _smartErrorService.saveError(errorDto);
  //   // });

  //   return wallets;
  // }

  // Future<String> createNewWallet(String walletName, String ownerId) async {
  //   Timestamp creationDate = Timestamp.fromDate(DateTime.now());

  //   var batch = dbReference.batch();

  //   var docRef = dbReference.collection(_walletsCollectionName).doc();

  //   String id = docRef.id;

  //   List<String> membersId = [ownerId];

  //   batch.set(docRef, {
  //     'id': id,
  //     'name': walletName,
  //     'ownerId': ownerId,
  //     'membersId': membersId,
  //     'creationDate': creationDate,
  //   });

  //   await batch.commit().then((value) {
  //     return id;
  //   }).catchError((onError) {
  //     SmartError errorDto = SmartError();
  //     errorDto.errorData = onError;
  //     errorDto.errorLog = onError.toString();
  //     errorDto.feature = 'Create new wallet';
  //     errorDto.userId = id;
  //     errorDto.userMail = '';

  //     _smartErrorService.saveError(errorDto);
  //     return '';
  //   });

  //   return id;
  // }
}

File: API/functions/src/db/wallet-dao.ts
import { Wallet } from "../models/wallet";
import { db } from "../index";

export class WalletDao {
    walletsCollectionName = Wallet.collectionName

    async getWalletsByUserId(userId: string): Promise<Wallet[]> {
        let walletList: Wallet[] = []

        var snapshot = await db
            .collection(this.walletsCollectionName)
            .where('membersId', 'array-contains', userId)
            .get()

        for (const doc of snapshot.docs) {
            let wallet = new Wallet(doc)
            walletList.push(wallet)
        }

        return walletList
    }

    async getById(walletId: string): Promise<Wallet | null> {
        let wallet: Wallet | null = null

        var snapshot = await db
            .collection(this.walletsCollectionName)
            .doc(walletId)
            .get()

        if (snapshot.exists) {
            wallet = new Wallet(snapshot)
        }

        return wallet
    }

    async getWalletByNameAndOwnerId(walletName: string, userId: string): Promise<Wallet | null> {
        let wallet: Wallet | null = null

        var snapshot = await db
            .collection(this.walletsCollectionName)
            .where('ownerId', '==', userId)
            .where('name', '==', walletName)
            .get()

        if (!snapshot.empty) {
            for (const doc of snapshot.docs) {
                wallet = new Wallet(doc)
            }
        }

        return wallet
    }

    async updateMemberIdList(members: string[], wallet: Wallet) {
        const batch = db.batch()

        var walletId = wallet.id

        const walletDocRef = db.collection(this.walletsCollectionName).doc(walletId)

        let now = new Date()

        batch.set(walletDocRef, {
            membersId: members,
            lastUpdate: now,
        }, { merge: true })

        await batch.commit()

        return walletId
    }

    async createNewWallet(userId: string, walletName: string) {
        let now = new Date()

        const batch = db.batch()

        const walletDocRef = db.collection(this.walletsCollectionName).doc()

        var uid = walletDocRef.id

        let membersId: string[] = [userId];

        batch.set(walletDocRef, {
            id: uid,
            name: walletName,
            ownerId: userId,
            membersId: membersId,
            creationDate: now,
        })

        await batch.commit()

        return uid
    }
}